scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,1016 @@
+        Maximum number of times to subdivide the region of integration in order to
+        improve the estimate over a region with high error.
+    args : tuple, optional
+        Additional positional args passed to ``f``. See Examples.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy.integrate._cubature import *
+    >>> def f(x, alphas):
+    ...     # Example function to integrate
+    ...     # f(x) = exp(sum(alpha * x))
+    ...
+    ...     # Support arbitrary shaped x and alphas
+    ...     ndim = x.shape[0]
+    ...     num_eval_points = x.shape[-1]
+    ...
+    ...     alphas_reshaped = np.expand_dims(alphas, -1)
+    ...     x_reshaped = x.reshape(
+    ...         ndim, *([1]*(len(alphas.shape) - 1)), num_eval_points
+    ...     )
+    ...
+    ...     return np.exp(np.sum(alphas_reshaped * x_reshaped, axis=0))
+    >>> rule_1d = GaussKronrod(21) # Use Gauss-Kronrod, which has error estimate
+    >>> alphas_1d = np.array([0.5])
+    >>> res = cub(
+    ...     f,
+    ...     np.array([0]),
+    ...     np.array([1]),
+    ...     rule_1d,
+    ...     args=(alphas_1d,)
+    ... ); res.estimate
+     np.float64(1.2974425414002562)
+    >>> # Now calculate 3D integral, constructing 3D rule from product of 1D rules
+    >>> product_rule = Product([rule_1d, rule_1d, rule_1d])
+    >>> # New values of alpha, now f calcualtes exp(0.2*x_1 + 0.5*x_2 + 0.8*x_3)
+    >>> alphas_3d = np.array([0.2, 0.5, 0.8])
+    >>> res = cub(
+    ...     f,
+    ...     np.array([0, 0, 0]),
+    ...     np.array([1, 1, 1]),
+    ...     product_rule,
+    ...     args=(alphas_3d,)
+    ... ); res.estimate
+     np.float64(2.2002853017758057)
+    >>> # Evaluate for many values of alpha simultaneously
+    >>> alphas_many = np.array([[0.1, 0.2, 2], [0.25, 0.5, 5], [0.4, 0.8, 8]])
+    >>> res = cub(
+    ...     f,
+    ...     np.array([0, 0, 0]),
+    ...     np.array([1, 1, 1]),
+    ...     rule_3d,
+    ...     args=(alphas_many,)
+    ... ); res.estimate
+     array([1.46914007e+00, 2.20028530e+00, 3.50827109e+04])
+    """
+
+    if max_subdivisions is None:
+        max_subdivisions = np.inf
+
+    est = rule.estimate(f, a, b, args, kwargs)
+
+    try:
+        err = rule.error_estimate(f, a, b, args, kwargs)
+    except NotImplementedError:
+        raise Exception("attempting cubature with a rule that doesn't implement error \
+estimation.")
+
+    regions = [CubatureRegion(est, err, a, b)]
+    subdivisions = 0
+    success = False
+
+    while np.any(err > atol + rtol * np.abs(est)):
+        # region_k is the region with highest estimated error
+        region_k = heapq.heappop(regions)
+
+        est_k = region_k.estimate
+        err_k = region_k.error
+
+        a_k, b_k = region_k.a, region_k.b
+
+        # Subtract the estimate of the integral and its error from the current global
+        # estimates, since these will be refined in the loop over all subregions.
+        est -= est_k
+        err -= err_k
```

Comment:
Are we concerned with error accumulation? It's an actual question, to which I don't know the answer. Probably OK to push it to later.